changelog/cpp98.dd

```diff
@@ -0,0 +1,57 @@
+## Transition to C++11 character types
+
+        With C++11 comes the advent of changed character type mangling.
+        D's default behavior is now to conform to this. However, this
+        will break existing code. Therefore, the old behavior can be
+        retained by using the -version=Cpp98 compiler switch.
+        For Win32 compilations using Digital Mars C++, -version=Cpp98 will
+        be predefined.
+
+        Of particular note is the new difference between wchar and wchar_t on
+        Windows. This will manifest itself as compile errors when
+        interfacing wchar* with wchar_t* code when calling the Windows API.
+        A cast will resolve the issue.
+
+        Going forward we recommend using WCHAR instead of wchar or wchar_t
+        when interfacing to Windows API functions. (WCHAR is Microsoft
+        Windows' 16 bit character type.)
+
+### C++ Type Equivalence
+
+#### Cpp98 behavior:
+
+        D       Posix           DMC++ Windows   VC++ Windows
+
+        wchar   unsigned short  wchar_t         wchar_t
+        dchar   wchar_t         unsigned        unsigned
+        wchar_t wchar_t         wchar_t         wchar_t
+        WCHAR   --              wchar_t         wchar_t
```

Comment:
I think "old behaviour" is a better term to use than "Cpp98" behaviour, which feels misleading...
If we were implementing C++98 behaviour, shouldn't it look like this:
```
        D       Posix           DMC++ Windows   VC++ Windows
        wchar           ** ERROR: CAN'T MANGLE **
        dchar           ** ERROR: CAN'T MANGLE **
        wchar_t wchar_t         wchar_t         wchar_t
        WCHAR   --              wchar_t         wchar_t
```
?
There's one additional detail though; VC++ has a compiler option where you can choose to NOT mangle `wchar_t` and instead it is a typedef for `unsigned short`... I've seen code compiled that way, but not for over a decade. I recommend not making any concession for that deprecated case.

I guess the reason for that hack dates back to when people used C compilers with no mangling distinction, and in C, `WCHAR` is a typedef for `unsigned short`, so some C programmers used `unsigned short` directly in their code, and hence the option in VC++ appeared to mangle for compatibility with those old programs.

Reply:
I think Manu is saying that D's `wchar` and `dchar` do not exist in C++98, which is why strictly speaking the table does not describe C++98 behavior, just the way DMD approached it.

Reply:
> But it can and does mangle them for C++98, per the chart.

But it's wrong, and it's always been wrong. There is no mangling for `wchar` or `dchar` in C++.
I'm just saying that calling is "C++98" is misleading.

Reply:
`wchar_t` is a keyword in C++98, and a distinct type, and has its own mangling. This is reflected in the chart, where `wchar` uses the `wchar_t` mangling for Windows, and `dchar` uses the `wchar_t` mangling for Posix.

Reply:
> where wchar uses the wchar_t mangling for Windows, and dchar uses the wchar_t mangling for Posix.

`wchar` and `dchar` have no equivalent type in C++. I think it's improper to map them to `wchar_t`. I understand it's for backwards compatibility with legacy DMD behaviour, but what I'm saying is, I don't think it's correct to call that mapping "Cpp98", because that's not what it is... it's just a mapping that DMD implemented.

I would expect a mode called "C++98 compatibility" to issue a compile error on account of no counterpart type in the C++98 language. Same way we give compile errors for `T[]`, etc.

Reply:
> wchar and dchar have no equivalent type in C++98.

Neither do any of the D basic types, not even `int`. We just look for the most pragmatic mapping.